omnichain/staking/contracts/Staking.sol

```diff
@@ -29,74 +31,132 @@ contract Staking is ERC20, zContract {
         updateRewards(staker);
     }
 
     function updateRewards(address staker) internal {
-        uint256 timeDifference = block.timestamp - lastStakeTime[staker];
-        uint256 rewardAmount = timeDifference * stakes[staker] * rewardRate;
+        uint256 rewardAmount = queryRewards(staker);
 
         _mint(beneficiaries[staker], rewardAmount);
         lastStakeTime[staker] = block.timestamp;
     }
 
-    function claimRewards(address staker) external {
-        if (beneficiaries[staker] != msg.sender) {
-            revert NotAuthorizedToClaim();
-        }
+    function unstakeZRC(address staker) internal {
+        uint256 amount = stakes[staker];
 
         updateRewards(staker);
-    }
-
-    function unstakeZRC(uint256 amount) external {
-        updateRewards(msg.sender);
-
-        require(stakes[msg.sender] >= amount, "Insufficient staked balance");
 
         address zrc20 = systemContract.gasCoinZRC20ByChainId(chainID);
+        (, uint256 gasFee) = IZRC20(zrc20).withdrawGasFee();
+
+        require(amount >= gasFee, "Amount should be greater than the gas fee");
 
-        (address gasZRC20, uint256 gasFee) = IZRC20(zrc20).withdrawGasFee();
+        bytes memory recipient = withdrawAddresses[staker];
 
         IZRC20(zrc20).approve(zrc20, gasFee);
-        IZRC20(zrc20).withdraw(abi.encodePacked(msg.sender), amount - gasFee);
+        IZRC20(zrc20).withdraw(recipient, amount - gasFee);
+
+        stakes[staker] = 0;
+        require(stakes[staker] <= amount, "Underflow detected");
+
+        lastStakeTime[staker] = block.timestamp;
+    }
+
+    function setBeneficiary(address staker, bytes calldata message) internal {
+        address beneficiary;
+        if (chainID == BITCOIN) {
+            beneficiary = BytesHelperLib.bytesToAddress(message, 1);
+        } else {
+            (, beneficiary) = abi.decode(message, (uint8, address));
+        }
+        beneficiaries[staker] = beneficiary;
+    }
 
-        stakes[msg.sender] -= amount;
-        lastStakeTime[msg.sender] = block.timestamp;
+    function setWithdraw(
+        address staker,
+        bytes calldata message,
+        bytes memory origin
+    ) internal {
+        bytes memory withdraw;
+        if (chainID == BITCOIN) {
+            withdraw = bytesToBech32Bytes(message, 1);
+        } else {
+            withdraw = origin;
+        }
+        withdrawAddresses[staker] = withdraw;
     }
 
-    function queryRewards(address account) public view returns (uint256) {
-        uint256 timeDifference = block.timestamp - lastStakeTime[account];
-        uint256 rewardAmount = timeDifference * stakes[account] * rewardRate;
+    function queryRewards(address staker) public view returns (uint256) {
+        uint256 timeDifference = block.timestamp - lastStakeTime[staker];
+        uint256 rewardAmount = timeDifference * stakes[staker] * rewardRate;
         return rewardAmount;
     }
+
+    function claimRewards(address staker) public {
```

Comment:
```suggestion
    function claimRewards(address beneficiary) public {
```

Reply:
We're using the `staker` address (or the Bitcoin equivalent) as a key to get the value from the `beneficiaries` mapping, so I'm not sure about this change.

Reply:
Ok, I thought it was indexed by beneficiary and it was a typo